rollout/trafficrouting.go

```diff
@@ -17,152 +17,177 @@ import (
-		return nginx.NewReconciler(nginx.ReconcilerConfig{
+		trafficReconcilers = append(trafficReconcilers, nginx.NewReconciler(nginx.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.kubeclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 			IngressLister:  c.ingressesLister,
-		}), nil
+		}))
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.ALB != nil {
-		return alb.NewReconciler(alb.ReconcilerConfig{
+		alb_reconcilier, err := alb.NewReconciler(alb.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.kubeclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 			IngressLister:  c.ingressesLister,
 		})
+		trafficReconcilers = append(trafficReconcilers, alb_reconcilier)
+		if err != nil {
+			return trafficReconcilers, err
+		}
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.SMI != nil {
-		return smi.NewReconciler(smi.ReconcilerConfig{
+		smi_reconcilier, err := smi.NewReconciler(smi.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.smiclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 		})
+		trafficReconcilers = append(trafficReconcilers, smi_reconcilier)
+		if err != nil {
+			return trafficReconcilers, err
+		}
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.Ambassador != nil {
 		ac := ambassador.NewDynamicClient(c.dynamicclientset, rollout.GetNamespace())
-		return ambassador.NewReconciler(rollout, ac, c.recorder), nil
+		trafficReconcilers = append(trafficReconcilers, ambassador.NewReconciler(rollout, ac, c.recorder))
 	}
+
+	// ensure that the trafficReconcilers is a healthy list and its not empty
+	if len(trafficReconcilers) > 0 {
+		return trafficReconcilers, nil
+	}
+
 	return nil, nil
 }
 
 func (c *rolloutContext) reconcileTrafficRouting() error {
-	reconciler, err := c.newTrafficRoutingReconciler(c)
+	reconcilers, err := c.newTrafficRoutingReconciler(c)
+	// a return here does ensure that all trafficReconcilers are healthy
+	// and sane in syntax
 	if err != nil {
 		return err
 	}
-	if reconciler == nil {
+	// ensure that trafficReconcilers list is healthy
+	if len(reconcilers) == 0 {
+		c.log.Info("No TrafficRouting Reconcilers found")
 		return nil
 	}
-	c.log.Infof("Reconciling TrafficRouting with type '%s'", reconciler.Type())
+	c.log.Infof("Found %d TrafficRouting Reconcilers", len(reconcilers))
+	// iterate over the list of trafficReconcilers
+	for _, reconciler := range reconcilers {
+		c.log.Infof("Reconciling TrafficRouting with type '%s'", reconciler.Type())
 
-	var canaryHash, stableHash string
-	if c.stableRS != nil {
-		stableHash = c.stableRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]
-	}
-	if c.newRS != nil {
-		canaryHash = c.newRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]
-	}
-	err = reconciler.UpdateHash(canaryHash, stableHash)
-	if err != nil {
-		return err
-	}
+		var canaryHash, stableHash string
```

Comment:
can you fix indentations? as it is unclear which of them changed.. this block isnt changed

Reply:
this is some visual effect of github diff.
open the file in ur IDE, you won't see any indentation issue.
Line 97 adds a loop, where we have to iterate over the list of trafficReconcilers.
indent is right